contracts/interchain-token-service/src/abi.rs

```diff
@@ -0,0 +1,464 @@
+use alloy_primitives::{FixedBytes, Uint, U256};
+use alloy_sol_types::{sol, SolValue};
+use axelar_soroban_std::ensure;
+use soroban_sdk::{contracterror, Bytes, BytesN, Env, String};
+
+use crate::types::{self, HubMessage, Message};
+extern crate alloc;
+use crate::abi::alloc::{string::String as StdString, vec};
+
+#[contracterror]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[repr(u32)]
+pub enum MessageError {
+    InsufficientMessageLength = 0,
+    InvalidMessageType = 1,
+    AbiDecodeFailed = 2,
+    InvalidAmount = 3,
+}
+
+sol! {
+    enum MessageType {
+        InterchainTransfer,
+        DeployInterchainToken,
+        DeployTokenManager, // note, this case is not supported by the ITS hub
+        SendToHub,
+        ReceiveFromHub
+    }
+
+    struct InterchainTransfer {
+        uint256 messageType;
+        bytes32 tokenId;
+        bytes sourceAddress;
+        bytes destinationAddress;
+        uint256 amount;
+        bytes data;
+    }
+
+    struct DeployInterchainToken {
+        uint256 messageType;
+        bytes32 tokenId;
+        string name;
+        string symbol;
+        uint8 decimals;
+        bytes minter;
+    }
+
+    struct SendToHub {
+        uint256 messageType;
+        string destination_chain;
+        bytes message;
+    }
+
+    struct ReceiveFromHub {
+        uint256 messageType;
+        string source_chain;
+        bytes message;
+    }
+}
+
+impl Message {
+    pub fn abi_encode(self, env: &Env) -> Bytes {
+        let msg = match self {
+            Self::InterchainTransfer(types::InterchainTransfer {
+                token_id,
+                source_address,
+                destination_address,
+                amount,
+                data,
+            }) => InterchainTransfer {
+                messageType: MessageType::InterchainTransfer.into(),
+                tokenId: FixedBytes::<32>::new(token_id.into()),
+                sourceAddress: source_address.to_alloc_vec().into(),
+                destinationAddress: destination_address.to_alloc_vec().into(),
+                amount: amount.try_into().expect("failed to convert"),
+                data: into_vec(data).into(),
+            }
+            .abi_encode_params(),
+            Self::DeployInterchainToken(types::DeployInterchainToken {
+                token_id,
+                name,
+                symbol,
+                decimals,
+                minter,
+            }) => DeployInterchainToken {
+                messageType: MessageType::DeployInterchainToken.into(),
+                tokenId: FixedBytes::<32>::new(token_id.into()),
+                name: to_std_string(name),
+                symbol: to_std_string(symbol),
+                decimals: decimals.try_into().expect("failed to convert"),
+                minter: into_vec(minter).into(),
+            }
+            .abi_encode_params(),
+        };
+        Bytes::from_slice(env, &msg)
```

Comment:
Similar to `to_alloc_vec`, can we ask Stellar team to add a `from_alloc_vec` for completeness?